### ISSUE TITLE
Input file prepare

### ISSUE THREAD

**Issue by zunyun-Gong** [jkwang93/Token-Mol#14]
Thank you for sharing this excellent work!

I would like to use the pre-trained Token-Mol model to extract molecular representations for downstream analysis. However, I am unsure how to prepare the input files (specifically, the CSV files in data/caco2/seed66/*.csv) from SMILES strings. Could you please provide detailed guidance on preparing these input files? Thank you very much!

**Comment by zunyun-Gong** [comment]
I have some problems for the calculation of torsion angles. For each SMILES, I obtained the same number of torsion angles as you did, but their values differ. Below is the code used for the calculation.

`
```
def GetDihedral(conf, atom_idx):
  return rdMolTransforms.GetDihedralRad(conf, atom_idx[0], atom_idx[1], atom_idx[2], atom_idx[3])

def get_torsion_angles(mol):
    torsions_list = []
    G = nx.Graph()
    for i, atom in enumerate(mol.GetAtoms()):
        G.add_node(i)
    nodes = set(G.nodes())
    for bond in mol.GetBonds():
        start, end = bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()
        G.add_edge(start, end)
    for e in G.edges():
        G2 = copy.deepcopy(G)
        G2.remove_edge(*e)
        if nx.is_connected(G2): continue
        l = list(sorted(nx.connected_components(G2), key=len)[0])
        if len(l) < 2: continue
        n0 = list(G2.neighbors(e[0]))
        n1 = list(G2.neighbors(e[1]))
        torsions_list.append(
            (n0[0], e[0], e[1], n1[0])
        )
    return torsions_list
token_mol_smiles = 'N C C [C@H] ( N ) C ( = O ) N 1 C C C C C 1'
token_mol_smiles = "".join(token_mol_smiles.split(" "))
mol_old = Chem.MolFromSmiles(token_mol_smiles)
params = AllChem.ETKDGv3()
params.randomSeed = 51
AllChem.EmbedMolecule(mol_old, params)
AllChem.UFFOptimizeMolecule(mol_old)

atom_idx = get_torsion_angles(mol_old)

conf = mol_old.GetConformer()
torsion_list = []
for i in atom_idx:
torsion_list.append(GetDihedral(conf,i))
print(torsion_list)`
```
The results are [1.2370299335939494, -2.8817789606971367, -2.0229567790843923, -0.035027643610659076], while the torsion angel from your ./data/val_mol_input.pkl are 'N C C [C@H] ( N ) C ( = O ) N 1 C C C C C 1 GEO 2.34 0.37 -0.69 -3.06'. Can you provides some details for the calculation of torsion angles.``

**Comment by sorui-qin** [comment]
Hi, thank you for your interest in our work. 

We represent torsion angles in radians, within the range of −3.14 to 3.14, retaining two decimal places. Torsion angles expressed in radians are numerically very sensitive, and it is therefore unsurprising that conformations minimized with the UFF force field do not exactly match the post-docking conformations from CrossDocked2020.

Below are the conformations aligned using obrms: the green conformation corresponds to the validation-set sequence, reconstructed using the `confs_to_mols_pocket.py` procedure, while the blue conformation is generated by the code you provided. The minimized heavy-atom RMSD between the two is **1.3 Å**. Although there are differences between the conformations, they can still be considered to be **within an acceptable range**.

<img width="2340" height="1135" alt="Image" src="https://github.com/user-attachments/assets/6c1dcba9-b237-4bbc-91a0-db89861357ea" />

Regarding the implementation details related to torsion angles, most of the relevant functions are located in `utils/standardlization.py`, which corresponds to the code you referenced. We welcome further discussion with you on this topic.

Rui Qin

**Comment by zunyun-Gong** [comment]
Thank you very much for the detailed reply! I noticed that your team used conformations from the GEOM dataset for pretraining. Does this choice help minimize the uncertainty and errors typically introduced by classical force fields?

Furthermore, regarding the post-docking conformations from the CrossDocked2020 dataset you referenced, it appears that protein-ligand docking is essential for obtaining accurate torsion angles of ligands in their protein-bound states. In contrast, deriving torsions directly from SMILES strings would only capture free (unbound) ligand conformations, which often differ significantly due to protein-induced fits.

Given that our task involves computing ligand representations by Token-Mol pre-training model to serve as features for downstream analysis, should we extract torsion angles from the docked complex data rather than SMILES strings? Would this approach be more suitable for integration with Token-Mol?

**Comment by sorui-qin** [comment]
Thanks for your further questions, I will respond point by point:

- The primary motivation for pretraining with GEOM is to leverage high-quality conformations at DFT-level accuracy, which are more precise than those obtained from traditional force fields, in order to support the other two downstream tasks that involve three-dimensional conformations and to improve the model’s capability in conformation generation. Another consideration concerns the trade-off between data quality and the computational cost imposed by dataset scale during pretraining. GEOM represents a dataset with sufficiently high quality and an appropriate scale for this purpose.

- The CrossDocked2020 dataset is used only during fine-tuning for the pocket-based molecular generation task. Therefore, any conformation-related inductive bias introduced by protein–ligand complexes affects only this single downstream task, and does not impact other downstream tasks. 

- You mentioned applying Token-Mol for representation learning, which we assume corresponds to tasks such as property prediction. We conducted preliminary experiments prior to this work (not disclosed in the paper), and the results indicated that, for identical SMILES strings, different conformations do not lead to *statistically significant differences* in prediction outcomes. This observation is consistent with the fact that these properties are generally insensitive to molecular conformation.  

- We carried out different conformations of the same molecule in RDKit via **MMFF force-field optimization** in the test above (which also addresses your initial question). Thus, in property prediction tasks, we still incorporate reasonably optimized conformations obtained through the MMFF force field to maintain consistency in the task paradigm, while also ensuring that the input sequences remain within the scope supported by our preliminary findings (i.e., we verified that properties do not vary significantly across reasonable MMFF-optimized conformations, but we do not make claims regarding abnormal conformations). Consequently, you can directly obtain torsion angle information from SMILES for representation learning.

**Comment by zunyun-Gong** [comment]
I am truly grateful for the thorough explanation you provided, which has completely resolved my concerns. Thank you for your meticulous work on this project, as well as the personal time and attention you gave to my questions.  I am now confident moving forward.  Looking forward to more excellent work from your team.
